### ISSUE TITLE
Janusgraph hadoop deserializes a wrong edgeID

### ISSUE THREAD

**Issue by ministat** [JanusGraph/janusgraph#3367]
When I use CloneVertexProgram to export a graph by Spark, my customer found there are duplicate edgeID for different edges. I investigated this and found the edgeID is different from what is queried from the backend storage.

The RCA is JansuGraphVertexDeserializer.java uses StarGraph (it used TinkerGraph before the fix https://github.com/JanusGraph/janusgraph/pull/3346) which generates a long edge ID by itself. The TinkerGraph also has a self maintained IDManager. But that is not the EdgeID queried through OLTP gremlin.

An example of this:

the edge of the format "srcID => edgeID => dstID": 819264 =>12965384 =>49272
But if I get the edgeID by gremlin against the backend directly, I got the edgeID is "7pw5k-hk5c-36d-120o" which is generated by RelationIdentifier. while, the EdgeID in the json is "12965384".

A single long variable as edgeID is difficult to be unique globally, but the RelationIdentifier can be unique because it combines the 4 IDs together: outVertexID + relationTypeID + relationID + inVertexID

The deserialization process needs to patch the edge ID through RelationIdentifier to be consistent with OLTP.

```
gremlin> graph = JanusGraphFactory.open('conf/fdb-spark.properties')
==>standardjanusgraph[org.janusgraph.diskstorage.foundationdb.FoundationDBStoreManager:[127.0.0.1]]
gremlin> g = graph.traversal()
==>graphtraversalsource[standardjanusgraph[org.janusgraph.diskstorage.foundationdb.FoundationDBStoreManager:[127.0.0.1]], standard]
gremlin> g.V(819264).outE().where(inV().has(T.id, 49272));
==>e[7pw5k-hk5c-36d-120o][819264-works->49272]
gremlin> 
```
The json extracted from exported graph looks like:
```
{
    "id": {
        "@type": "g:Int64",
        "@value": 819264
    },
    "label": "employee",
    "inE": {
        "knows": [
            ...
     },
     "outE": {
        "works": [
            {
                "id": {
                    "@type": "g:Int64",
                    "@value": 12965384
                },
                "inV": {
                    "@type": "g:Int64",
                    "@value": 49272
                },
                "properties": {
                    "year": {
                        "@type": "g:Int32",
                        "@value": 12
                    },
                    "position": "PM"
                }
            },
....

```
Please include configurations and logs if available.

For confirmed bugs, please report:

- Version: 0.6.2
- Storage Backend: cql, hbase
- Mixed Index Backend: 
- Link to discussed bug: 
- Expected Behavior:
- Current Behavior:
- Steps to Reproduce: See the above description
### Stack Trace (if you have one)

```
paste stack trace here
```

**Comment by li-boxuan** [comment]
> the edge of the format "srcID => edgeID => dstID": 819264 =>12965384 =>49272
But if I get the edgeID by gremlin against the backend directly, I got the edgeID is "7pw5k-hk5c-36d-120o" which is generated by RelationIdentifier. while, the EdgeID in the json is "12965384".

`12965384` is ought to be a unique value across the whole database. FWIW `7pw5k == LongEncoding.encode(12965384)`.

> my customer found there are duplicate edgeID for different edges

This seems to indicate there is either ID corruption, or there was some bug when ids were being generated.

In short, the current edgeId should already be unique, and if it's not, it's another bug, not the hadoop deserializer's bug.

Aside from the duplicate ID issue (which is very suspicious but it seems to be another issue), did you see any other problem with the current approach?

**Comment by ministat** [comment]
> This seems to indicate there is either ID corruption, or there was some bug when ids were being generated.

That is interesting! It seems that the edgeID "12965384" is the RelationID, but I checked the default edgeID generation for both StarGraph and TinkerGraph, I found both of them generate ID through an atomic increasement:
https://github.com/apache/tinkerpop/blob/13663c10a2ec1e1eca975f8745946e5a79acb36d/gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/structure/util/star/StarGraph.java#L92, https://github.com/apache/tinkerpop/blob/13663c10a2ec1e1eca975f8745946e5a79acb36d/tinkergraph-gremlin/src/main/java/org/apache/tinkerpop/gremlin/tinkergraph/structure/TinkerGraph.java#L694
But for JanusGraphEdge, the ID genereation is https://github.com/JanusGraph/janusgraph/blob/master/janusgraph-core/src/main/java/org/janusgraph/graphdb/transaction/StandardJanusGraphTx.java#L806:L808. 
When I debugging I found the edgeID created by StandardJanusGraphTx is "relationId-outVertexId-typeId-inVertexId", that is why I'm trying to do the same on JanusgraphVertexDeserializer.

I also guess the RelationID is not unique in my test, there are 10132 duplications for a graph with 15 billion vertice and 23 billion edges, but I have not yet understand how to guarantee the uniqueness of RelationID. Does Janusgraph have any document describing the ID manager?

>Aside from the duplicate ID issue (which is very suspicious but it seems to be another issue), did you see any other problem with the current approach?

I did not see other problem yet. But I'm a little worried about the index system built on the IDs if I run some index query.

**Comment by li-boxuan** [comment]
>  JanusGraphEdge, the ID generation is https://github.com/JanusGraph/janusgraph/blob/master/janusgraph-core/src/main/java/org/janusgraph/graphdb/transaction/StandardJanusGraphTx.java#L806:L808.

That's the correct entry point to start looking into! Unfortunately, there's not much doc about ID manager except Javadoc.

> When I debugging I found the edgeID created by StandardJanusGraphTx is "relationId-outVertexId-typeId-inVertexId", that is why I'm trying to do the same on JanusgraphVertexDeserializer

Yeah, that makes sense. It's a breaking change but seems reasonable to me.

**Comment by SinyoWong** [comment]
Sir , do you know how to use CloneVertexProgram to import dataset to a cql graph by Spark ？
Could you please give me an example ? Thks a loooooot!
